packages/runtime/container-runtime/src/gc/garbageCollection.ts

```diff
@@ -537,30 +664,145 @@ export class GarbageCollector implements IGarbageCollector {
-			updatedGCData = this.runSweepPhase(sweepReadyNodes, gcData);
-		} else if (this.configs.testMode) {
-			// If we are running in GC test mode, delete objects for unused routes. This enables testing scenarios
-			// involving access to deleted data.
+		/**
+		 * Currently, there are 3 modes for sweep:
+		 * Test mode - Unreferenced nodes are immediately deleted without waiting for them to be sweep ready.
+		 * Tombstone mode - Sweep ready modes are marked as tombstones instead of being deleted.
+		 * Sweep mode - Sweep ready modes are deleted.
+		 *
+		 * These modes serve as staging for applications that want to enable sweep by providing an incremental
+		 * way to test and validate sweep works as expected.
+		 */
+		if (this.configs.testMode) {
+			// If we are running in GC test mode, unreferenced nodes (gcResult.deletedNodeIds) are deleted.
 			this.runtime.updateUnusedRoutes(gcResult.deletedNodeIds);
-		} else if (this.configs.tombstoneMode) {
+			return gcData;
+		}
+
+		if (this.configs.tombstoneMode) {
 			this.tombstones = sweepReadyNodes;
 			// If we are running in GC tombstone mode, update tombstoned routes. This enables testing scenarios
 			// involving access to "deleted" data without actually deleting the data from summaries.
-			// Note: we will not tombstone in test mode.
 			this.runtime.updateTombstonedRoutes(this.tombstones);
+			return gcData;
 		}
 
-		this.gcDataFromLastRun = cloneGCData(updatedGCData);
+		if (!this.configs.shouldRunSweep) {
+			return gcData;
+		}
 
-		// Log pending unreferenced events such as a node being used after inactive. This is done after GC runs and
-		// updates its state so that we don't send false positives based on intermediate state. For example, we may get
-		// reference to an unreferenced node from another unreferenced node which means the node wasn't revived.
-		await this.telemetryTracker.logPendingEvents(logger);
+		// 1. Call the runtime to delete sweep ready nodes. The runtime returns a list of nodes it deleted.
+		// TODO: GC:Validation - validate that removed routes are not double delete and that the child routes of
+		// removed routes are deleted as well.
+		const deletedNodeIds = this.runtime.deleteSweepReadyNodes(sweepReadyNodes);
 
-		return gcStats;
+		// 2. Clear unreferenced state tracking for deleted nodes.
+		for (const nodeId of deletedNodeIds) {
+			const nodeStateTracker = this.unreferencedNodesState.get(nodeId);
+			// TODO: GC:Validation - assert that the nodeStateTracker is defined
+			if (nodeStateTracker !== undefined) {
+				// Stop tracking so as to clear out any running timers.
+				nodeStateTracker.stopTracking();
+				// Delete the node as we don't need to track it any more.
+				this.unreferencedNodesState.delete(nodeId);
+			}
+			// TODO: GC:Validation - assert that the deleted node is not a duplicate
+			this.deletedNodes.add(nodeId);
+		}
+
+		// 3. Update and return the passed in gcData by removing deleted nodes from it.
```

Comment:
I'd rephrase this, you're not modifying the passed in gcData.  Also made me consider a slight refactor which I'm trying locally and will run by you if I like it

Reply:
Rephrased it.
I had initially deleted nodes from gcData directly (so had Tyler when he originally wrote this code). But it gives a tslint warning which made me think it's not safe to do so.